R/morphemepiece.R

```diff
@@ -126,67 +126,176 @@ magrittr::`%>%`
 
 # .mp_tokenize_word -----------------------------------------------------------
 
-#' Apply Morphemepiece tokenization to word
+#' Tokenize a Word
 #'
-#' Repeatedly pull out the longest valid vocabulary piece from the word until
-#' the whole word is represented in terms of vocabulary pieces. This algorithm
-#' is intended to be used as a "fall-back" for words that are not found in the
-#' morphemepiece lookup table.
+#' Tokenize a single "word" (no whitespace). The word can technically contain
+#' punctuation, but typically punctuation has been split off by this point.
 #'
-#' @param word Character; word to tokenize.
-#' @param vocab A morphemepiece vocabulary.
-#'
-#' @return A character vector corresponding to the original word, broken into 
-#' pieces.
+#' This is an adapdation of wordpiece:::.tokenize_word. The main differences are
+#' that it was designed to work with a morphemepiece vocabulary, which can
+#' include prefixes (denoted like "pre##"). As in wordpiece, the algorithm uses
+#' a repeated greedy search for the largest piece from the vocabulary found
+#' within the word, but starting from either the beginning or the end of the
+#' word (controlled by the `dir` parameter). The input vocabulary must be split
+#' into prefixes, suffixes, and "words".
+#' 
+#' @param word Word to tokenize.
+#' @param vocab_split List of named integer vector containing vocabulary words.
+#'   Should have components named "prefixes", "words", "suffixes"
+#' @param dir Integer; if 1 (the default), look for tokens starting at the
+#'   beginning of the word. Otherwise, start at the end.
+#' @param unk_token Token to represent unknown words.
+#' @param max_chars Maximum length of word recognized.
 #'
+#' @return Input word as a list of tokens.
 #' @keywords internal
-.mp_tokenize_word <- function(word, vocab) {
-    # handle prefixes a little differently...
-    prefixes <- vocab[stringr::str_ends(vocab, "##")]
-    names(prefixes) <- stringr::str_sub(prefixes, end = -3)
+.mp_tokenize_word <- function(word, 
+                              vocab_split, 
+                              dir = 1, # -1 for backwards
+                              unk_token = "[UNK]",
+                              max_chars = 100) {
+  if (stringi::stri_length(word) > max_chars) {
```

Comment:
`nchar` is 10x faster than `stringi::stri_length`. Do we care about the exception cases?

Reply:
nope, changing!